server/sonar-server/src/main/java/org/sonar/server/authentication/JwtHttpHandler.java

```diff
@@ -0,0 +1,174 @@
+public class JwtHttpHandler {
+
+  private static final String SESSION_TIMEOUT_PROPERTY = "sonar.authentication.timeoutInHours";
+  private static final int SESSION_TIMEOUT_DEFAULT_VALUE = 72;
+
+  private static final String JWT_COOKIE = "JWT-SESSION";
+  private static final String LAST_REFRESH_TIME_PARAM = "lastRefreshTime";
+
+  private static final String CSRF_JWT_PROPERTY = "xsrfToken";
+
+  // Time after which a user will be disconnected
+  private static final int SESSION_DISCONNECT_IN_SECONDS = 3 * 30 * 24 * 60 * 60;
+
+  // This refresh time is used to refresh the session
+  // The value must be lower than SESSION_TIMEOUT_IN_SECONDS
+  private static final int SESSION_REFRESH_IN_SECONDS = 5 * 60;
+
+  private static final String RAILS_USER_ID_SESSION = "user_id";
+
+  private final System2 system2;
+  private final DbClient dbClient;
+  private final Server server;
+  private final JwtSerializer jwtSerializer;
+
+  // This timeout is used to disconnect the user we he has not browse any page for a while
+  private final int sessionTimeoutInSeconds;
+  private final JwtCsrfVerifier jwtCsrfVerifier;
+
+  public JwtHttpHandler(System2 system2, DbClient dbClient, Server server, Settings settings, JwtSerializer jwtSerializer, JwtCsrfVerifier jwtCsrfVerifier) {
+    this.jwtSerializer = jwtSerializer;
+    this.server = server;
+    this.dbClient = dbClient;
+    this.system2 = system2;
+    this.sessionTimeoutInSeconds = getSessionTimeoutInHours(settings) * 60 * 60;
+    this.jwtCsrfVerifier = jwtCsrfVerifier;
+  }
+
+  void generateToken(String userLogin, HttpServletResponse response) {
+    String csrfState = jwtCsrfVerifier.generateState(response, sessionTimeoutInSeconds);
+
+    String token = jwtSerializer.encode(new JwtSerializer.JwtSession(
+      userLogin,
+      sessionTimeoutInSeconds,
+      ImmutableMap.of(
+        LAST_REFRESH_TIME_PARAM, system2.now(),
+        CSRF_JWT_PROPERTY, csrfState)));
+    response.addCookie(createCookie(JWT_COOKIE, token, sessionTimeoutInSeconds));
+  }
+
+  void validateToken(HttpServletRequest request, HttpServletResponse response) {
+    Optional<Cookie> jwtCookie = findCookie(JWT_COOKIE, request);
+    if (jwtCookie.isPresent()) {
+      Cookie cookie = jwtCookie.get();
+      String token = cookie.getValue();
+      if (!isNullOrEmpty(token)) {
+        validateToken(token, request, response);
+      }
+    }
+  }
+
+  private void validateToken(String tokenEncoded, HttpServletRequest request, HttpServletResponse response) {
+    Optional<Claims> claims = jwtSerializer.decode(tokenEncoded);
+    if (!claims.isPresent()) {
+      removeSession(request, response);
+      return;
+    }
+
+    Date now = new Date(system2.now());
+
+    Claims token = claims.get();
+    if (now.after(DateUtils.addSeconds(token.getIssuedAt(), SESSION_DISCONNECT_IN_SECONDS))) {
+      removeSession(request, response);
+      return;
+    }
+
+    Optional<UserDto> user = getUser(claims.get().getSubject());
```

Comment:
I recommend to rename `getUser` to something like `selectUserFromDb` or `loadUser`